Tests/SourceKitLSPTests/TypeHierarchyTests.swift

```diff
@@ -0,0 +1,167 @@
+        fatalError("Type hierarchy at \(testLoc) was empty")
+      }
+      let request = TypeHierarchySupertypesRequest(item: item)
+      let types = try ws.sk.sendSync(request)
+      return types ?? []
+    }
+
+    func subtypes(at testLoc: TestLocation) throws -> [TypeHierarchyItem] {
+      guard let item = try typeHierarchy(at: testLoc).first else {
+        fatalError("Type hierarchy at \(testLoc) was empty")
+      }
+      let request = TypeHierarchySubtypesRequest(item: item)
+      let types = try ws.sk.sendSync(request)
+      return types ?? []
+    }
+
+    // Convenience functions
+
+    func testLoc(_ name: String) -> TestLocation {
+      ws.testLoc(name)
+    }  
+
+    func loc(_ name: String) -> Location {
+      Location(badUTF16: ws.testLoc(name))
+    }
+
+    func withoutData(_ item: TypeHierarchyItem) -> TypeHierarchyItem {
+      var item = item
+      item.data = nil
+      return item
+    }
+
+    /// Compares the given type hierarchies ignoring the implementation-specific
+    /// data field (which includes e.g. USRs that are difficult to test, especially
+    /// in the presence of extensions, and are not user-visible anyway).
+    func assertEqualIgnoringData(
+      _ actual: [TypeHierarchyItem],
+      _ expected: [TypeHierarchyItem],
+      file: StaticString = #file,
+      line: UInt = #line
+    ) {
+      XCTAssertEqual(
+        actual.map(withoutData),
+        expected.map(withoutData),
+        file: file,
+        line: line
+      )
+    }
+
+    func item(_ name: String, _ kind: SymbolKind, detail: String = "main", at locName: String) -> TypeHierarchyItem {
+      let location = loc(locName)
+      return TypeHierarchyItem(
+        name: name,
+        kind: kind,
+        tags: nil,
+        detail: detail,
+        uri: location.uri,
+        range: location.range,
+        selectionRange: location.range
+      )
+    }
+
+    // Test type hierarchy preparation
+
+    assertEqualIgnoringData(try typeHierarchy(at: testLoc("P")), [
+      item("P", .interface, at: "P"),
+    ])
+    assertEqualIgnoringData(try typeHierarchy(at: testLoc("A")), [
+      item("A", .class, at: "A"),
+    ])
+    assertEqualIgnoringData(try typeHierarchy(at: testLoc("S")), [
+      item("S", .struct, at: "S"),
+    ])
+    assertEqualIgnoringData(try typeHierarchy(at: testLoc("E")), [
+      item("E", .enum, at: "E"),
+    ])
+
+    // Test supertype hierarchy
+
+    assertEqualIgnoringData(try supertypes(at: testLoc("A")), [])
+    assertEqualIgnoringData(try supertypes(at: testLoc("B")), [
+      item("A", .class, at: "A"),
+      item("P", .interface, at: "P"),
+    ])
+    assertEqualIgnoringData(try supertypes(at: testLoc("C")), [
+      item("B", .class, at: "B"),
+    ])
+    assertEqualIgnoringData(try supertypes(at: testLoc("D")), [
+      item("A", .class, at: "A"),
+    ])
+    assertEqualIgnoringData(try supertypes(at: testLoc("S")), [
+      item("P", .interface, at: "P"),
+    ])
```

Comment:
The supertype hierarchy doesn't seem to include the protocol `X`, which `P` retroactively conforms to:

```swift
protocol P {}
protocol X {}
struct S: P {}
extension S: X {}
```

This is due to the index treating the extension `S: X` as a separate entity (it is also listed as such when enumerating the subtypes of `X`).

Reply:
Could we work around this by doing an index query of extensions for `S` first and also including all the extension’s supertypes?

I think as far as subtypes are concerned, it’s fine to show the extension. I don’t have strong feelings about that.

Reply:
Yeah, I think we could do that, we would probably have do an index query per extension though (which could be a few).

Reply:
I don’t expect these queries to take very long, so I think it would be worth it.